openwisp_controller/subnet_division/handlers.py

```diff
@@ -0,0 +1,111 @@
+import logging
+
+from django.db import transaction
+from django.db.models.signals import post_delete, post_save, pre_save
+from django.dispatch import receiver
+from swapper import load_model
+
+from . import tasks
+
+logger = logging.getLogger(__name__)
+
+Subnet = load_model('openwisp_ipam', 'Subnet')
+IpAddress = load_model('openwisp_ipam', 'IpAddress')
+SubnetDivisionRule = load_model('subnet_division', 'SubnetDivisionRule')
+SubnetDivisionIndex = load_model('subnet_division', 'SubnetDivisionIndex')
+VpnClient = load_model('config', 'VpnClient')
+
+
+_subnet_division_rule_update_queue = dict()
+# It is used to monitor changes in fields of a SubnetDivisionRule object
+# An entry is added to the queue from pre_save signal in the following format
+#
+# '<rule-uid>: {
+#   '<field-name>': '<old-value>',
+# }
+#
+# In post_save signal, it is checked whether entry for SubnetDivisionRule object
+# exists in this queue. If it exists changes are made to related objects.
+
+
+@receiver(
+    pre_save,
+    sender=SubnetDivisionRule,
+    dispatch_uid='subnet_division_rule_changed_fields',
+)
+def subnet_division_rule_changed_fields(sender, instance, **kwargs):
+    try:
+        db_instance = SubnetDivisionRule.objects.get(id=instance.id)
+    except SubnetDivisionRule.DoesNotExist:
+        # This rule does not exists in database.
+        # No operation is needed to be performed.
+        return
+    else:
+        # Check which fields of instance is modified
+        # NOTE: Open-ended implementation to allow change in all
+        # fields of SubnetDivisionRule in future.
+        # Currently only changing label and number of IPs is allowed.
+        modified_fields = {}
+        for field in db_instance._meta.fields:
+            instance_value = getattr(instance, field.name)
+            db_value = getattr(db_instance, field.name)
+            if instance_value != db_value:
+                modified_fields[field.name] = db_value
+
+        _subnet_division_rule_update_queue[str(instance.id)] = modified_fields
```

Comment:
Let's move this logic to the models, following the structure of the existing code of OpenWISP Controller, if you look in the other apps, you will see signal receivers and celery tasks always call methods on the model, this allows to concentrate the core logic in one place which is good, because using signals usually makes the code harder to read.

It also gives another advantage: the method can be overridden by apps which extend openwisp controller and provide customized models.